CSP 2358 provider courses patch endpoint to  update has online delivery option

raviudari01 opened the pull request.

raviudari01 committed:
CSP-2358 courses patch end point to update HasOnlineDeliveryOption

raviudari01 committed:
CSP-2358 Updated PatchProviderCourseCommandHandler and ProviderCoursesWriteRepository to allow the HasOnlineDeliveryOption property to be patched

raviudari01 committed:
CSP-2358 Added tests for patch command edge cases and validation

raviudari01 committed:
CSP-2358  Expanded unit tests for PatchProviderCourseCommand to cover code coverage

raviudari01 committed:
CSP-2358 Refactor  code change classes to use filebased namespace

DevenKShah commented on src/SFA.DAS.Roatp.Application.UnitTests/ProviderCourse/Commands/PatchProviderCourse/PatchProviderCourseCommandHandlerTests.cs:
```diff
@@ -11,63 +11,64 @@
+        patchCommand.Replace(path => path.ContactUsPhoneNumber, patch.ContactUsPhoneNumber);
+        patchCommand.Replace(path => path.StandardInfoUrl, patch.StandardInfoUrl);
+        patchCommand.Replace(path => path.IsApprovedByRegulator, patch.IsApprovedByRegulator);
+        patchCommand.Replace(path => path.HasOnlineDeliveryOption, patch.HasOnlineDeliveryOption);
```
Since a new field was introduced, I would write a separate test for this field. Ideally I would like to see a test for each field being patched.

DevenKShah commented on src/SFA.DAS.Roatp.Application.UnitTests/ProviderCourse/Commands/PatchProviderCourse/PatchProviderCourseCommandTests.cs:
```diff
@@ -4,119 +4,297 @@
+        };
+
+        command.HasOnlineDeliveryOption.Should().BeNull();
+        command.IsPresentHasOnlineDeliveryOption.Should().BeTrue();
```
You cannot unset the value or set it to null, so I am not able to follow this test.

DevenKShah commented on src/SFA.DAS.Roatp.Application/ProviderCourse/Commands/PatchProviderCourse/PatchProviderCourseCommandValidator.cs:
```diff
@@ -6,59 +6,65 @@
+        RuleFor(c => c.HasOnlineDeliveryOption != null)
+            .Equal(true)
+            .When(c => c.IsPresentHasOnlineDeliveryOption)
+            .WithMessage(HasOnlineDeliveryOptionIsNotABooleanErrorMessage);
```
```suggestion
        RuleFor(c => c.HasOnlineDeliveryOption)
            .NotEmpty()
            .When(c => c.IsPresentHasOnlineDeliveryOption)
            .WithMessage(HasOnlineDeliveryOptionIsNotABooleanErrorMessage);
```

DevenKShah commented on src/SFA.DAS.Roatp.Application.UnitTests/Providers/Commands/PatchProvider/PatchProviderCommandTests.cs:
These tests are completely wrong. They should be targeting PatchProviderCommand, instead they are targeting PatchProviderCourseCommand. Classic copy paste and forget error. We need to fix this.

DevenKShah commented on src/SFA.DAS.Roatp.Application.UnitTests/ProviderCourse/Queries/ProviderAllCoursesQueryHandlerTests.cs:
```diff
@@ -13,159 +13,161 @@
+            course.LarsCode = larsCodeIncrement.ToString();
+            course.Standard = new Standard { LarsCode = course.LarsCode, IsRegulatedForProvider = false };
+            course.IsApprovedByRegulator = true;
+            course.Locations = [new ProviderCourseLocation()];
```
not sure why these changes are made?

raviudari01 committed:
CSP-2358 Refactored unit tests to target PatchProviderCommand and PatchProvider model.

raviudari01 committed:
Update src/SFA.DAS.Roatp.Application/ProviderCourse/Commands/PatchProviderCourse/PatchProviderCourseCommandValidator.cs

Co-authored-by: Deven Shah <devenshah1974@gmail.com>

raviudari01 committed:
CSP-2358 Renamed the parameter isNoErrorExpected to isValid in the
Validate_Patch_HasOnlineDeliveryOption_VariousFieldValues_MatchingErrors
test method

raviudari01 commented on src/SFA.DAS.Roatp.Application.UnitTests/ProviderCourse/Commands/PatchProviderCourse/PatchProviderCourseCommandValidatorTests.cs:
```diff
@@ -11,454 +11,525 @@
+    [TestCase("false", true)]
+    [TestCase("not boolean", false)]
+    [TestCase("2", false)]
+    public async Task Validate_Patch_HasOnlineDeliveryOption_VariousFieldValues_MatchingErrors(string hasOnlineDeliveryOptionValue, bool isNoErrorExpected)
```
Done

raviudari01 committed:
Merge branch 'CSP-2358-Courses-Update-HasOnlineDeliveryOption' of https://github.com/SkillsFundingAgency/das-roatp-api into CSP-2358-Courses-Update-HasOnlineDeliveryOption

raviudari01 committed:
CSP-2358 Enhanced HasOnlineDeliveryOption parsing to accept both boolean and integer (1/0) values.

raviudari01 committed:
Revert "CSP-2358 Enhanced HasOnlineDeliveryOption parsing to accept both boolean and integer (1/0) values."

This reverts commit 4dc4c8979b4ce820e6113d8f36d709bdc7fa8763.

raviudari01 committed:
CSP-2358 resolved merge conflicts

raviudari01 committed:
CSP-2358 resolved merge conflicts

raviudari01 committed:
Merge branch 'main' into CSP-2358-Courses-Update-HasOnlineDeliveryOption

raviudari01 committed:
Merge branch 'main' into CSP-2358-Courses-Update-HasOnlineDeliveryOption
